table/server/master/src/main/java/alluxio/master/table/transform/TransformManager.java

```diff
@@ -249,55 +251,60 @@ private void completePartitionTransformation(String dbName, String tableName, Tr
       }
     }
 
-    private void removeTransformationJobForTable(Pair<String, String> table) {
-      mJobs.remove(table);
+    private void onFinish(TransformJobInfo job) {
+      mJobHistory.put(job.getJobId(), job);
       RemoveTransformJobEntry journalEntry = RemoveTransformJobEntry.newBuilder()
-          .setDbName(table.getFirst())
-          .setTableName(table.getSecond())
+          .setDbName(job.getDb())
+          .setTableName(job.getTable())
           .build();
       try (JournalContext journalContext = mCreateJournalContext.apply()) {
         applyAndJournal(journalContext, Journal.JournalEntry.newBuilder()
             .setRemoveTransformJob(journalEntry).build());
       } catch (UnavailableException e) {
         LOG.error("Failed to create journal for RemoveTransformJob for database {} table {}",
-            table.getFirst(), table.getSecond());
+            job.getDb(), job.getTable());
       }
     }
 
     /**
      * Handle the cases where a job fails, is cancelled, or job ID not found.
      *
-     * @param table the (database, table) pair
+     * @param job the transformation job
+     * @param status the job status
+     * @param error the job error message
      */
-    private void handleJobError(Pair<String, String> table) {
-      removeTransformationJobForTable(table);
+    private void handleJobError(TransformJobInfo job, Status status, String error) {
+      job.setJobStatus(status);
+      job.setJobErrorMessage(error);
+      onFinish(job);
     }
 
     /**
      * Handle the case where a job is completed.
      *
-     * @param table the (database, table) pair
      * @param job the transformation job
      */
-    private void handleJobSuccess(Pair<String, String> table, TransformJob job) {
+    private void handleJobSuccess(TransformJobInfo job) {
       try {
-        completePartitionTransformation(table.getFirst(), table.getSecond(), job);
+        updatePartition(job);
+        job.setJobStatus(Status.COMPLETED);
       } catch (IOException e) {
-        LOG.error("Failed to update partition layouts for database {} table {}",
-            table.getFirst(), table.getSecond());
+        String error = String.format("Failed to update partition layouts for database %s table %s",
+            job.getDb(), job.getTable());
+        LOG.error(error);
+        job.setJobStatus(Status.FAILED);
+        job.setJobErrorMessage(error);
       }
```

Comment:
would there be a problem if the server crashes between the partition update is journaled and the job completion is journaled?

Reply:
resolved offline